In [3]:
import pandas as pd

# Car Price Analysis - Guided Assignment

**Student Name:** _[Doreen Kendi]_

**Date:** _[03/07/2025]_

---

## Assignment Overview

**Business Scenario**: You are a data analyst working for a car dealership chain. The management wants to understand market trends, identify profitable segments, and optimize their inventory strategy.

**Your Task**: Apply pandas data manipulation techniques to analyze car pricing data and provide business recommendations.

**Learning Objectives**:
- Apply index and column operations
- Use loc and iloc for data selection
- Implement boolean masking and filtering
- Perform sorting and ranking operations
- Handle missing values appropriately
- Create and transform columns
- Conduct group-by aggregations
- Build pivot tables for analysis
- Generate business insights from data

**Dataset**: Car prices dataset with 558,837 records and 16 features

---

## Instructions
1. Complete each section by writing code in the provided cells
2. Follow the step-by-step guidance
3. Answer the questions in markdown cells
4. Run all cells to ensure your code works
5. Provide business insights based on your analysis

## Part 1: Data Loading and Initial Exploration (10 points)

### Task 1.1: Import Libraries
Import the necessary libraries for data analysis.

In [4]:
# TODO: Import pandas, numpy, and warnings
# Hint: Use standard aliases (pd, np)
# TODO: Filter warnings to keep output clean

# Your code here:
import pandas as pd
import numpy as np
import warnings

### Task 1.2: Load the Dataset
Load the car prices dataset and display basic information.

In [5]:
# TODO: Load the 'car_prices.csv' dataset
# TODO: Print the dataset shape
# TODO: Display the first 5 rows

# Your code here:
car_prices = pd.read_csv(r"C:\Users\muriu\OneDrive\Desktop\DS ASSIGNMENTS\car_prices.csv")
car_prices.shape
car_prices.head(5)
print(f"Dataset shape: {car_prices.shape}")


Dataset shape: (558837, 16)


**Question 1.1**: What is the shape of the dataset? How many cars and features are there?

**Answer**: _[558837,16]_

## Part 2: Index and Column Operations (15 points)

### Task 2.1: Examine and Clean Column Names
Display the column names and clean them for consistency.

In [6]:
# TODO: Print original column names
# TODO: Clean column names (remove spaces, convert to lowercase)
# TODO: Print cleaned column names

# Your code here:
print("Original columns:")
print(car_prices.columns.tolist())
car_prices.columns= car_prices.columns.str.strip().str.replace(' ', '_')
print("Cleaned columns:")
print(car_prices.columns.tolist())

Original columns:
['year', 'make', 'model', 'trim', 'body', 'transmission', 'vin', 'state', 'condition', 'odometer', 'color', 'interior', 'seller', 'mmr', 'sellingprice', 'saledate']
Cleaned columns:
['year', 'make', 'model', 'trim', 'body', 'transmission', 'vin', 'state', 'condition', 'odometer', 'color', 'interior', 'seller', 'mmr', 'sellingprice', 'saledate']


### Task 2.2: Data Type Exploration
Use the `.info()` method to understand the data types and missing values.

In [7]:
# TODO: Display dataset info using .info()
# TODO: Display the first 3 rows using .head(3)

# Your code here:
car_prices.info()
car_prices.isnull().sum()
car_prices.head(3)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 558837 entries, 0 to 558836
Data columns (total 16 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   year          558837 non-null  int64  
 1   make          548536 non-null  object 
 2   model         548438 non-null  object 
 3   trim          548186 non-null  object 
 4   body          545642 non-null  object 
 5   transmission  493485 non-null  object 
 6   vin           558833 non-null  object 
 7   state         558837 non-null  object 
 8   condition     547017 non-null  float64
 9   odometer      558743 non-null  float64
 10  color         558088 non-null  object 
 11  interior      558088 non-null  object 
 12  seller        558837 non-null  object 
 13  mmr           558799 non-null  float64
 14  sellingprice  558825 non-null  float64
 15  saledate      558825 non-null  object 
dtypes: float64(4), int64(1), object(11)
memory usage: 68.2+ MB


year make     model        trim   body transmission                vin  \
0  2015  Kia   Sorento          LX    SUV    automatic  5xyktca69fg566472   
1  2015  Kia   Sorento          LX    SUV    automatic  5xyktca69fg561319   
2  2014  BMW  3 Series  328i SULEV  Sedan    automatic  wba3c1c51ek116351   

  state  condition  odometer  color interior  \
0    ca        5.0   16639.0  white    black   
1    ca        5.0    9393.0  white    beige   
2    ca       45.0    1331.0   gray    black   

                                   seller      mmr  sellingprice  \
0                 kia motors america  inc  20500.0       21500.0   
1                 kia motors america  inc  20800.0       21500.0   
2  financial services remarketing (lease)  31900.0       30000.0   

                                  saledate  
0  Tue Dec 16 2014 12:30:00 GMT-0800 (PST)  
1  Tue Dec 16 2014 12:30:00 GMT-0800 (PST)  
2  Thu Jan 15 2015 04:30:00 GMT-0800 (PST)

**Question 2.1**: Which column represents the selling price of cars? What data type is it?

**Answer**: _[sellingprice, float]_

## Part 3: Data Selection with Loc and Iloc (15 points)

### Task 3.1: Using Loc for Label-based Selection
Practice selecting data using row and column labels.

In [8]:
# TODO: Select first 10 rows and columns 'make', 'model', 'year', 'sellingprice'
# Use .loc for this selection

# Your code here:
car_prices.loc[:9, ['make','model','year','sellingprice']]

make                model  year  sellingprice
0        Kia              Sorento  2015       21500.0
1        Kia              Sorento  2015       21500.0
2        BMW             3 Series  2014       30000.0
3      Volvo                  S60  2015       27750.0
4        BMW  6 Series Gran Coupe  2014       67000.0
5     Nissan               Altima  2015       10900.0
6        BMW                   M5  2014       65000.0
7  Chevrolet                Cruze  2014        9800.0
8       Audi                   A4  2014       32250.0
9  Chevrolet               Camaro  2014       17500.0

### Task 3.2: Using Iloc for Position-based Selection
Practice selecting data using integer positions.

In [9]:
# TODO: Select first 5 rows and first 6 columns using .iloc
# TODO: Display the result

# Your code here:
car_prices.iloc[0:5,0:6]

year   make                model        trim   body transmission
0  2015    Kia              Sorento          LX    SUV    automatic
1  2015    Kia              Sorento          LX    SUV    automatic
2  2014    BMW             3 Series  328i SULEV  Sedan    automatic
3  2015  Volvo                  S60          T5  Sedan    automatic
4  2014    BMW  6 Series Gran Coupe        650i  Sedan    automatic

## Part 4: Boolean Masking and Filtering (20 points)

### Task 4.1: Explore Categorical Data
Examine the unique values in categorical columns.

In [10]:
# TODO: Get all categorical columns (object dtype)
# TODO: For the first 3 categorical columns, display:
#       - Column name and number of unique values
#       - First 10 unique values

# Your code here:
categorical_columns= car_prices.select_dtypes(include =["object"]).columns
print("Categorical Columns and their unique values:")
for col in categorical_columns[:3]:
  print(f"\n{col}: {car_prices[col].nunique()} unique values")
  print(f"Sample values: {car_prices[col].unique()[:10]}")



Categorical Columns and their unique values:

make: 96 unique values
Sample values: ['Kia' 'BMW' 'Volvo' 'Nissan' 'Chevrolet' 'Audi' 'Ford' 'Hyundai' 'Buick'
 'Cadillac']

model: 973 unique values
Sample values: ['Sorento' '3 Series' 'S60' '6 Series Gran Coupe' 'Altima' 'M5' 'Cruze'
 'A4' 'Camaro' 'A6']

trim: 1963 unique values
Sample values: ['LX' '328i SULEV' 'T5' '650i' '2.5 S' 'Base' '1LT'
 '2.0T Premium Plus quattro' 'LT' '3.0T Prestige quattro']


### Task 4.2: Price-based Filtering
Create filters based on price ranges.

In [11]:
# TODO: Calculate price percentiles (25th, 75th, 90th)
# TODO: Print the percentile values with proper formatting
# TODO: Create filters for:
#       - Budget cars (bottom 25%)
#       - Luxury cars (top 25%)
# TODO: Print the count of cars in each category

# Your code here:
price_col = 'sellingprice'
price_25th = car_prices[price_col].quantile(0.25)
price_75th = car_prices[price_col].quantile(0.75)
price_90th = car_prices[price_col].quantile(0.90)
print(f"Price Statistics:")
print(f"25th percentile: ${price_25th:,.2f}")
print(f"75th percentile: ${price_75th:,.2f}")
print(f"90th percentile: ${price_90th:,.2f}")

luxury_cars = car_prices[car_prices[price_col]>=price_75th]
print(f"\nLuxury cars (top 25% by price): {len(luxury_cars)} cars")

price_25th = car_prices[price_col].quantile(0.25)
budget_cars = car_prices[car_prices[price_col] <= price_25th]
print(f"\nbudget cars (bottom 25% by price): {len(budget_cars)} cars")




Price Statistics:
25th percentile: $6,900.00
75th percentile: $18,200.00
90th percentile: $25,500.00

Luxury cars (top 25% by price): 140682 cars

budget cars (bottom 25% by price): 140853 cars


### Task 4.3: Complex Filtering
Apply multiple conditions to find specific car segments.

In [16]:
# TODO: Find cars that are:
#       - Luxury price (top 25%) AND
#       - Made by BMW, Mercedes, or Audi
# Hint: Use .str.contains() with the 'or' operator (|)
# TODO: Display the count and first 5 rows of this filtered data

# Your code here:
price_75th = car_prices['sellingprice'].quantile(0.75)
luxury_cars = car_prices[car_prices['sellingprice'] >= price_75th]
luxury_brands = luxury_cars[
    luxury_cars['make'].str.lower().str.contains('bmw|mercedes|audi', na=False)
]
print(f"\nLuxury cars (BMW, Mercedes, or Audi): {len(luxury_brands)} cars")
print(luxury_brands.head())
total_cars = len(car_prices)
luxury_cars_count = len(luxury_cars)
luxury_percentage = (luxury_cars_count / total_cars) * 100
print(f"Luxury cars represent {luxury_percentage:.2f}% of the total dataset.")





Luxury cars (BMW, Mercedes, or Audi): 24566 cars
    year  make                model                       trim   body  \
2   2014   BMW             3 Series                 328i SULEV  Sedan   
4   2014   BMW  6 Series Gran Coupe                       650i  Sedan   
6   2014   BMW                   M5                       Base  Sedan   
8   2014  Audi                   A4  2.0T Premium Plus quattro  Sedan   
10  2014  Audi                   A6      3.0T Prestige quattro  Sedan   

   transmission                vin state  condition  odometer  color interior  \
2     automatic  wba3c1c51ek116351    ca       45.0    1331.0   gray    black   
4     automatic  wba6b2c57ed129731    ca       43.0    2641.0   gray    black   
6     automatic  wbsfv9c51ed593089    ca       34.0   14943.0  black    black   
8     automatic  wauffafl3en030343    ca       42.0    9557.0  white    black   
10    automatic  wauhgafc0en062916    ca       48.0   14414.0  black    black   

                        

**Question 4.1**: What percentage of the total dataset do luxury cars represent?

**Answer**: _[25.17%]_

## Part 5: Sorting and Ranking (15 points)

### Task 5.1: Basic Sorting
Sort the data to find extreme values.

In [18]:
# TODO: Create a clean dataset by removing rows with missing values
# TODO: Sort by 'sellingprice' (descending) to find most expensive cars
# TODO: Display top 10 most expensive cars with columns: make, model, year, sellingprice
# TODO: Sort by 'sellingprice' (ascending) to find least expensive cars
# TODO: Display top 10 least expensive cars

# Your code here:
print("Missing values analysis:")
missing_data = car_prices.isnull().sum().sort_values(ascending=False)
missing_percentage = (car_prices.isnull().sum() / len(car_prices) * 100).sort_values(ascending=False)

missing_summary = pd.DataFrame({
    'Missing_Count': missing_data,
    'Missing_Percentage': missing_percentage
})

print(missing_summary[missing_summary['Missing_Count'] > 0])
df_analysis = car_prices.dropna().copy()
price_col = 'sellingprice'
most_expensive = df_analysis.sort_values(by=price_col, ascending=False).head(10)
least_expensive = df_analysis.sort_values(by=price_col, ascending=True).head(10)
print("TOP 10 MOST EXPENSIVE CARS:")
display_cols = [col for col in ['make', 'model', 'year', price_col] if col in df_analysis.columns]
print(most_expensive[display_cols])

print("\nTOP 10 LEAST EXPENSIVE CARS:")
print(least_expensive[display_cols])

Missing values analysis:
              Missing_Count  Missing_Percentage
transmission          65352           11.694287
body                  13195            2.361154
condition             11820            2.115107
trim                  10651            1.905922
model                 10399            1.860829
make                  10301            1.843292
color                   749            0.134028
interior                749            0.134028
odometer                 94            0.016821
mmr                      38            0.006800
sellingprice             12            0.002147
saledate                 12            0.002147
vin                       4            0.000716
TOP 10 MOST EXPENSIVE CARS:
                 make            model  year  sellingprice
344905           Ford           Escape  2014      230000.0
548169        Ferrari       458 Italia  2011      183000.0
446949  Mercedes-Benz          S-Class  2015      173000.0
545523    Rolls-Royce            Ghost 

### Task 5.2: Multi-column Sorting and Ranking
Perform more complex sorting operations.

In [22]:
# TODO: Sort data by 'make' (A-Z) and then by 'sellingprice' (high to low)
# TODO: Display first 15 rows with columns: make, model, sellingprice
# TODO: Add a 'price_rank' column using .rank() method
# TODO: Add a 'price_percentile' column using .rank(pct=True)
# TODO: Display the first 5 rows with ranking columns

# Your code here:
if 'make' in df_analysis.columns:
    sorted_cars = df_analysis.sort_values(by=['make', price_col], ascending=[True, False])
    
    print("Cars sorted by Make (A-Z) and Price (High to Low):")
    print(sorted_cars[['make', 'model', price_col]].head(15))
    df_analysis['price_rank'] = df_analysis[price_col].rank(method='dense', ascending=False)
df_analysis['price_percentile'] = df_analysis[price_col].rank(pct=True)

print("\nPrice rankings added:")
ranking_cols = [col for col in ['make', 'model', price_col, 'price_rank', 'price_percentile'] if col in df_analysis.columns]
min_price = car_prices['sellingprice'].min()
max_price = car_prices['sellingprice'].max()

print(f"Price range: ${min_price:,.2f} to ${max_price:,.2f}")
print(df_analysis[ranking_cols].head())

Cars sorted by Make (A-Z) and Price (High to Low):
         make model  sellingprice
282957  Acura   MDX       47000.0
341817  Acura   MDX       46500.0
61333   Acura   MDX       43800.0
69274   Acura   MDX       43800.0
6464    Acura   MDX       43250.0
50607   Acura   MDX       43250.0
107706  Acura   MDX       43250.0
41571   Acura   MDX       43000.0
369781  Acura   MDX       42400.0
69270   Acura   MDX       42100.0
78806   Acura   MDX       42000.0
188449  Acura   MDX       42000.0
107715  Acura   MDX       41750.0
240519  Acura   MDX       41750.0
503908  Acura   MDX       41750.0

Price rankings added:
Price range: $1.00 to $230,000.00
    make                model  sellingprice  price_rank  price_percentile
0    Kia              Sorento       21500.0      1033.0          0.832391
1    Kia              Sorento       21500.0      1033.0          0.832391
2    BMW             3 Series       30000.0       779.0          0.946915
3  Volvo                  S60       27750.0       83

**Question 5.1**: What is the price range (min to max) of the cars in the dataset?

**Answer**: _[$1.00 to $230,000.00]_

## Part 6: Missing Values Analysis (10 points)

### Task 6.1: Identify Missing Values
Analyze the pattern of missing values in the dataset.

In [24]:
# TODO: Calculate missing values count for each column
# TODO: Calculate missing values percentage for each column
# TODO: Create a summary DataFrame with columns: 'Missing_Count', 'Missing_Percentage'
# TODO: Display only columns that have missing values

# Your code here:
print("Missing values analysis:")
missing_data = car_prices.isnull().sum().sort_values(ascending=False)
missing_percentage = (car_prices.isnull().sum() / len(car_prices) * 100).sort_values(ascending=False)

missing_summary = pd.DataFrame({
    'Missing_Count': missing_data,
    'Missing_Percentage': missing_percentage
})

print(missing_summary[missing_summary['Missing_Count'] > 0])


Missing values analysis:
              Missing_Count  Missing_Percentage
transmission          65352           11.694287
body                  13195            2.361154
condition             11820            2.115107
trim                  10651            1.905922
model                 10399            1.860829
make                  10301            1.843292
color                   749            0.134028
interior                749            0.134028
odometer                 94            0.016821
mmr                      38            0.006800
sellingprice             12            0.002147
saledate                 12            0.002147
vin                       4            0.000716


**Question 6.1**: Which columns have the most missing values? What might be the business reason for this?

**Answer**: _[transmission]_

## Part 7: Creating and Transforming Columns (20 points)

### Task 7.1: Create Price Categories
Segment cars into price categories for business analysis.

In [27]:
# TODO: Create a function to categorize prices:
#       - Budget: < $10,000
#       - Mid-Range: $10,000 - $24,999
#       - Premium: $25,000 - $49,999
#       - Luxury: $50,000+
#       - Unknown: for missing values
# TODO: Apply this function to create 'price_category' column
# TODO: Display the count of cars in each category

# Your code here:
car_prices_business = car_prices.copy()
price_col = 'sellingprice'

def categorize_price(price):
    if pd.isna(price):
        return 'Unknown'
    elif price < 10000:
        return 'Budget'
    elif price < 25000:
        return 'Mid-Range'
    elif price < 50000:
        return 'Premium'
    else:
        return 'Luxury'

car_prices_business['price_category'] = car_prices_business[price_col].apply(categorize_price)
print(car_prices_business['price_category'].value_counts())


Mid-Range    286579
Budget       212258
Premium       56450
Luxury         3538
Unknown          12
Name: price_category, dtype: int64


### Task 7.2: Create Age-based Features
Calculate car age and create age categories.

In [30]:
# TODO: Calculate car age (current year 2024 minus car year)
# TODO: Create age categories:
#       - New: 0-3 years
#       - Recent: 4-8 years
#       - Older: 9+ years
# TODO: Display the count of cars in each age category

# Your code here:
current_year = 2024

if 'year' in car_prices_business.columns:
    # Calculate car age
    car_prices_business['car_age'] = current_year - car_prices_business['year']
    
    # Categorize age
    car_prices_business['age_category'] = car_prices_business['car_age'].apply(
        lambda x: 'New' if x <= 3 else 'Recent' if x <= 8 else 'Older'
    )
    
    # Display count of each age category
    print(car_prices_business['age_category'].value_counts())




Older    558837
Name: age_category, dtype: int64


### Task 7.3: String Operations and Brand Analysis
Identify luxury brands using string operations.

In [32]:
# TODO: Create 'is_luxury_brand' column that identifies luxury brands
# Luxury brands: BMW, Mercedes, Audi, Lexus, Acura, Infiniti, Porsche, Jaguar, Land Rover, Volvo, Cadillac
# Hint: Use .str.contains() with case=False
# TODO: Calculate and display the count and percentage of luxury brand cars

# Your code here:
if 'make' in car_prices.columns:
   car_prices['make_upper'] = car_prices['make'].str.upper()
   car_prices['is_luxury_brand'] = car_prices['make'].str.contains(
        'BMW|Mercedes|Audi|Lexus|Acura|Infiniti|Porsche|Jaguar|Land Rover|Volvo|Cadillac', case=False, na=False
    )
    
print(f"Luxury brand cars: {car_prices['is_luxury_brand'].sum()}")
print(f"Luxury brand percentage: {car_prices['is_luxury_brand'].mean()*100:.1f}%")

Luxury brand cars: 93197
Luxury brand percentage: 16.7%


**Question 7.1**: What percentage of cars in the dataset are from luxury brands?

**Answer**: _[16.7%]_

## Part 8: Group By Aggregations (25 points)

### Task 8.1: Price Analysis by Category
Analyze pricing patterns across different segments.

In [34]:
# TODO: Group by 'price_category' and calculate:
#       - count, mean, median, std, min, max for 'sellingprice'
# TODO: Round results to 2 decimal places
# TODO: Display the results
# TODO: Calculate market share (percentage) for each price category

# Your code here:
price_analysis = car_prices_business.groupby('price_category')[price_col].agg([
    'count', 'mean', 'median', 'std', 'min', 'max'
]).round(2)

print("PRICE ANALYSIS BY CATEGORY:")
print(price_analysis)
market_share = car_prices_business['price_category'].value_counts(normalize=True) * 100
print("\nMARKET SHARE BY PRICE CATEGORY:")
for category, share in market_share.items():
    print(f"{category}: {share:.1f}%")

PRICE ANALYSIS BY CATEGORY:
                 count      mean   median       std      min       max
price_category                                                        
Budget          212258   5175.53   5100.0   2895.76      1.0    9990.0
Luxury            3538  65026.14  58550.0  18003.94  50000.0  230000.0
Mid-Range       286579  15713.53  14900.0   4040.10  10000.0   24999.0
Premium          56450  31436.50  29600.0   5730.30  25000.0   49900.0
Unknown              0       NaN      NaN       NaN      NaN       NaN

MARKET SHARE BY PRICE CATEGORY:
Mid-Range: 51.3%
Budget: 38.0%
Premium: 10.1%
Luxury: 0.6%
Unknown: 0.0%


### Task 8.2: Brand Analysis
Identify top-performing brands by volume and price.

In [ ]:
# TODO: Group by 'make' and calculate:
#       - count, mean, median of 'sellingprice'
#       - most common price_category for each make
# TODO: Sort by average price (descending)
# TODO: Display top 15 makes by average price

# Your code here:
price_analysis =car_prices_business.groupby('make')[price_col].agg([
    'count', 'mean', 'median', 'std', 'min', 'max'
]).round(2)
price_category_mode = car_prices_business.groupby('make')['price_category'] \
 .agg(lambda x: x.mode().iloc[0])
price_analysis['most_common_category'] = price_category_mode
price_analysis_sorted = price_analysis.sort_values(by='mean', ascending=False)
print(price_analysis_sorted.head(15))


              count       mean    median       std       min       max  \
make                                                                     
Rolls-Royce      17  153488.24  149800.0  10490.88  140000.0  171500.0   
Ferrari          19  127210.53  124000.0  23109.23   81000.0  183000.0   
Lamborghini       4  112625.00  113000.0   4956.06  107000.0  117500.0   
Bentley         116   74367.67   62250.0  31933.93   22700.0  163000.0   
airstream         1   71000.00   71000.0       NaN   71000.0   71000.0   
Tesla            23   67054.35   66500.0   9675.99   49250.0   85750.0   
Aston Martin     25   54812.00   51000.0  14949.54   38300.0  103000.0   
Fisker            9   46461.11   46000.0   4891.43   37000.0   54500.0   
Maserati        133   45320.30   45250.0  22986.42    1200.0   95000.0   
Lotus             1   40800.00   40800.0       NaN   40800.0   40800.0   
Porsche        1383   39965.33   41500.0  24884.46    1250.0  157000.0   
Land Rover     1735   33097.86   33000

### Task 8.3: Luxury vs Non-Luxury Analysis
Compare luxury and non-luxury brands.

In [37]:
# TODO: Group by 'is_luxury_brand' and calculate:
#       - count, mean, median of 'sellingprice'
# TODO: Display results with proper labels (Non-Luxury, Luxury)
# TODO: Calculate the price premium of luxury brands (as percentage)

# Your code here:
luxury_brands = ['bmw', 'mercedes', 'audi']
car_prices_business['is_luxury_brand'] = car_prices_business['make'].str.lower().isin(luxury_brands)
luxury_group = car_prices_business.groupby('is_luxury_brand')['sellingprice'].agg(['count', 'mean', 'median']).round(2)
luxury_group.index = ['Non-Luxury' if not x else 'Luxury' for x in luxury_group.index]
print(luxury_group)
non_luxury_mean = luxury_group.loc['Non-Luxury', 'mean']
luxury_mean = luxury_group.loc['Luxury', 'mean']

price_premium_pct = ((luxury_mean - non_luxury_mean) / non_luxury_mean) * 100
print(f"\nLuxury brand price premium: {price_premium_pct:.2f}%")


             count      mean   median
Non-Luxury  532085  13235.33  12000.0
Luxury       26740  21093.81  19750.0

Luxury brand price premium: 59.38%


**Question 8.1**: Which price category has the largest market share? What does this tell us about the market?

**Answer**: _[Luxury]_

**Question 8.2**: What premium do luxury brands command over non-luxury brands?

**Answer**: _[59.38%]_

## Part 9: Pivot Tables (15 points)

### Task 9.1: Make vs Price Category Analysis
Create a comprehensive cross-tabulation.

In [38]:
# TODO: Get top 10 makes by volume
# TODO: Filter data to include only these top makes
# TODO: Create a pivot table with:
#       - index: 'make'
#       - columns: 'price_category'
#       - values: 'sellingprice'
#       - aggfunc: ['count', 'mean']
# TODO: Fill missing values with 0
# TODO: Round to 2 decimal places and display

# Your code here:
if 'make' in car_prices_business.columns and len(car_prices_business['make'].unique()) > 5:
  top_makes = car_prices_business['make'].value_counts().head(10).index
  df_pivot = car_prices_business[car_prices_business['make'].isin(top_makes)].copy()
  pivot_make_price = pd.pivot_table(
        df_pivot,
        index='make',
        columns='price_category',
        values=price_col,
        aggfunc=['count', 'mean'],
        fill_value=0
    ).round(2)
    
  print("PIVOT TABLE: Make vs Price Category")
  print(pivot_make_price)
else:
    print("Insufficient make data for pivot analysis")

PIVOT TABLE: Make vs Price Category
                count                                      mean            \
price_category Budget Luxury Mid-Range Premium Unknown   Budget    Luxury   
make                                                                        
BMW              5606    752      7128    7233       0  5621.44  62740.27   
Chevrolet       27996    327     27192    4682       0  5364.87  54938.53   
Chrysler         8243      0      8636     395       0  4908.69      0.00   
Dodge           14181      1     15429    1097       0  5604.79  51700.00   
Ford            29294    122     53415   10722       0  5457.93  55576.23   
Honda           11830      0     14548     828       0  4813.04      0.00   
Hyundai          7589      4     13885     333       0  5753.48  52100.00   
Kia              5763      0     12172     140       0  5608.29      0.00   
Nissan          17523     61     35670     692       0  5743.78  73405.74   
Toyota          15399     24     22285  

**Question 9.1**: Looking at the pivot table, which brands are most represented in the luxury segment?

**Answer**: _BMW, Chevrolet]_

## Part 10: Business Insights and Recommendations (25 points)

### Task 10.1: Generate Market Overview
Provide a comprehensive summary of the market.

In [39]:
# TODO: Calculate and display:
#       - Total cars analyzed
#       - Average price
#       - Median price
#       - Price range (min to max)
# TODO: Format all prices with proper currency formatting

# Your code here:
# Total cars
total_cars = len(car_prices_business)

# Price stats
avg_price = car_prices_business['sellingprice'].mean()
median_price = car_prices_business['sellingprice'].median()
min_price = car_prices_business['sellingprice'].min()
max_price = car_prices_business['sellingprice'].max()

# Display formatted summary
print(f"Total cars analyzed: {total_cars}")
print(f"Average price: ${avg_price:,.2f}")
print(f"Median price: ${median_price:,.2f}")
print(f"Price range: ${min_price:,.2f} to ${max_price:,.2f}")


Total cars analyzed: 558837
Average price: $13,611.36
Median price: $12,100.00
Price range: $1.00 to $230,000.00


### Task 10.2: Identify Top Volume Brands
Find the most popular brands by volume.

In [40]:
# TODO: Get top 5 brands by volume
# TODO: For each brand, display:
#       - Brand name
#       - Number of cars
#       - Average price
# TODO: Format the output nicely

# Your code here:
# Group by brand ('make'), and count and average price
brand_summary = car_prices_business.groupby('make')['sellingprice'].agg(['count', 'mean'])

# Sort by count (volume) descending
top_5_brands = brand_summary.sort_values(by='count', ascending=False).head(5)

# Round and format average price
top_5_brands['mean'] = top_5_brands['mean'].round(2)

# Display nicely
print("Top 5 Brands by Volume:\n")
for brand, row in top_5_brands.iterrows():
    print(f"- {brand}: {row['count']} cars, Average Price: ${row['mean']:,.2f}")


Top 5 Brands by Volume:

- Ford: 93553.0 cars, Average Price: $14,540.47
- Chevrolet: 60197.0 cars, Average Price: $11,977.40
- Nissan: 53946.0 cars, Average Price: $11,739.02
- Toyota: 39871.0 cars, Average Price: $12,249.32
- Dodge: 30708.0 cars, Average Price: $11,206.80


### Task 10.3: Business Recommendations
Based on your analysis, provide business recommendations.

**Question 10.1**: Based on your analysis, what are the key characteristics of the car market?

**Answer**: _[1.WMid-range and budget cars dominate the market: The majority of cars fall into the Budget (<$10,000) and Mid-Range ($10,000–$24,999) price categories, indicating strong demand in the affordable segment.
2.Luxury brands have a significant price premium: Brands like BMW, Mercedes, and Audi have an average selling price more than 100% higher than non-luxury brands.
3.Top brands by volume include Toyota, Ford, Honda, Nissan, and Chevrolet: These brands make up the bulk of the inventory and are generally associated with reliability and affordability.]_

**Question 10.2**: What recommendations would you give to a car dealership regarding inventory strategy?

**Answer**: _[1. Stock more mid-range and budget cars
2. Include selective luxury vehicle
3. Prioritize high-volume brands
4. Use pricing data to stay competitive]_

**Question 10.3**: Which segments offer the best opportunities for profit maximization?

**Answer**: _[1.Luxury cars offer the highest per-unit profit margin. Though they account for only ~25% of listings, their average price is more than double that of non-luxury brands. Strategic sales here can yield substantial profits, especially for low-mileage, well-maintained vehicles.
2.High-volume, reliable brands (e.g., Toyota, Honda) generate steady cash flow due to faster turnover rates. Stocking the best-selling models ensures predictable revenue.]_

## Bonus Section: Advanced Analysis (Extra Credit - 10 points)

### Bonus Task: Custom Analysis
Choose one additional analysis that would provide value to the business. Examples:
- Seasonal trends (if date data available)
- Geographic analysis by state
- Condition impact on pricing
- Mileage vs price relationships

Implement your chosen analysis below:

In [ ]:
# TODO: Implement your custom analysis here
# Be creative and provide business value!

# Your code here:

**Bonus Question**: Explain your custom analysis and the business insights it provides.

**Answer**: _[Explain your analysis and insights]_

---

## Assignment Checklist

Before submitting, ensure you have completed:

- [ ] All code cells run without errors
- [ ] All questions are answered
- [ ] Code is well-commented and readable
- [ ] Results are properly formatted
- [ ] Business insights are data-driven and actionable
- [ ] Student name and date are filled in at the top

**Total Points: 170 (+ 10 bonus)**

---

## Grading Rubric

| Component | Excellent (A) | Good (B) | Satisfactory (C) | Needs Improvement (D/F) |
|-----------|---------------|----------|------------------|-------------------------|
| **Code Quality** | Clean, efficient, well-commented | Good structure, minor issues | Works but could be cleaner | Multiple errors or poor structure |
| **Technical Skills** | Demonstrates mastery of all pandas concepts | Good use of most concepts | Basic understanding shown | Missing key concepts |
| **Analysis Quality** | Thorough, insightful analysis | Good analysis with minor gaps | Basic analysis completed | Superficial or incomplete |
| **Business Insights** | Clear, actionable recommendations | Good insights with data support | Basic insights provided | Weak or unsupported insights |
| **Communication** | Clear, professional presentation | Well-organized with minor issues | Adequately presented | Poor organization or clarity |

**Submission Instructions**: 
1. Ensure all cells are executed and outputs are visible
2. Save the notebook with your name: `Car_Analysis_[Your_Last_Name].ipynb`
3. Submit through the course management system
4. Include a brief reflection (2-3 sentences) on what you learned from this assignment